### PR TITLE
@constructedBy attribute

### PR DESCRIPTION
**Pull request by marek**
Added @constructedBy to document how objects are created that do not use the regular constructor.
For example:

HTMLAudioElement is created by new Audio()
HTMLVideoElement is created by document.createElement("video");

if a @constructedBy is specified, and no @constructor is present, the default constructor is not generated.
if both @constructedBy and @constructor are present. Both are generated

/**
- this is a MyClass description
- @class
- @constructor
- @param {Number} [foo] foo is a constructor param
- @constructedBy {document.createElement("MyClass")} {This is a description}
- // this is an example of how to create a new instance of MyClass
- var instance = document.createElement("MyClass");
  */

MyClass = function(foo) { }

### CODE REVIEW

File: build/bbPlugin.js [hunk 1]
```diff
@@ -33,7 +66,9 @@ JSDOC.PluginManager.registerPlugin(
                 var betaTag = symbol.comment.getTag("beta");
                 var paramCallbacks = symbol.comment.tags.filter(function($){return $.isCallback && $.title == "param"});
                 var fieldCBs = symbol.comment.tags.filter(function($){return $.isCallback && $.title == "field"});
-		var learnTag = symbol.comment.getTag("learns");
+                var learnTag = symbol.comment.getTag("learns");
+                var constructorTag = symbol.comment.getTag("constructor");
+                var constructedBy = symbol.comment.tags.filter(function($){return $.title=="constructedBy" && !!$.type});
```
Comment (jeffheifetz) [inline build/bbPlugin.js]: Is that a double negative?

File: build/bbPlugin.js [hunk 2]
```diff
@@ -126,9 +169,18 @@ JSDOC.PluginManager.registerPlugin(
...
-		}
+                if (docTag.title == "learns"){
+                    docTag.desc = docTag.nibbleName(docTag.desc);
+                }
+                if(docTag.title == "constructedBy") {
+                    if(!docTag.type) {
+                        LOG.warn(docTag.title + " missing type");
+                    } else {
+                        var parts = GetType(docTag.desc);
+                        docTag.desc = parts.type;
```
Comment (jeffheifetz) [inline build/bbPlugin.js:190]: The tag should already be parsed. You should not need to do so again

File: build/BBTemplate/class.tmpl [hunk 3]
```diff
@@ -436,6 +446,20 @@
 			</if>
 		</div>
 	</if>
+    {+ResetIndex()+}
```
Comment (jeffheifetz) [inline build/BBTemplate/class.tmpl:454]: I cannot make this call from BBPlugin. If you want to be able to you'll need to declare these in the publish.js

File: build/bbPlugin.js (hunk 1 above)
Comment (marek) [inline build/bbPlugin.js]: http://navirudra.com/blog/javascript/javascript-double-negative-trick.html

File: build/bbPlugin.js (hunk 2 above)
Comment (marek) [inline build/bbPlugin.js:190]: This parses two sets. jsDocs only parses one by default.

File: build/bbPlugin.js (hunk 1 above)
Comment (jeffheifetz) [inline build/bbPlugin.js]: Nice trick, Thanks! I never knew about this.
If it covers additional cases, should we throw this everywhere?

File: build/bbPlugin.js (hunk 2 above)
Comment (jeffheifetz) [inline build/bbPlugin.js:190]: Sorry didn't notice

File: build/BBTemplate/class.tmpl (hunk 3 above)
Comment (jeffheifetz) [inline build/BBTemplate/class.tmpl:454]: Got it working with our help, thanks

**Comment by jeffheifetz** [comment]
Seems like this should be good enough for now, we can always expand it in the future if we need different platform support per constructor

**Comment by jeffheifetz** [comment]
Is there a way to set different sets of parameters for each constructor?

**Comment by marek** [comment]
Nope there is no way of defining the parameters. 
Felt that it would be odd to describe the parameter types for a call like document.createElement("audio"); when it's a fixed string you use.

**Comment by jeffheifetz** [comment]
How will Jack use it for something like window.openDatabase that has like 4 parameters?

----- Original Message -----

**Comment by marek** [comment]
It's limited to just providing information in the description field. Or we can document 'window' more thoroughly

**Pull request merged**